lib/Conversion/SeqToSV/FirRegLowering.h

```diff
@@ -16,20 +16,121 @@
 #include "circt/Support/LLVM.h"
 #include "circt/Support/Namespace.h"
 #include "circt/Support/SymCache.h"
+#include "llvm/ADT/GraphTraits.h"
 
 namespace circt {
+namespace hw {
+namespace detail {
+
+struct SCCNode {
+  Operation *op;
+  llvm::function_ref<bool(Operation *)> filter;
+  explicit SCCNode(Operation *o, llvm::function_ref<bool(Operation *)> f)
+      : op(o), filter(f) {}
+  bool operator==(const SCCNode &other) const { return other.op == op; }
+  bool operator!=(const SCCNode &other) const { return !(*this == other); }
+};
+
+class SCCIterator
+    : public llvm::iterator_facade_base<SCCIterator, std::forward_iterator_tag,
+                                        SCCNode> {
+  SCCNode sccOp;
+  mlir::Value::use_iterator iterator, itEnd;
+  unsigned resultId, maxResultId;
+
+public:
+  explicit SCCIterator(SCCNode op, bool end = false) : sccOp(op) {
+    iterator = itEnd = mlir::Value::use_iterator();
+    if (sccOp.op->getNumResults() == 0) {
+      iterator = itEnd = mlir::Value::use_iterator();
+      return;
+    }
+    resultId = 0;
+    maxResultId = sccOp.op->getNumResults() - 1;
+    if (end)
+      iterator = itEnd;
+    else
+      iterator = sccOp.op->getResult(resultId).use_begin();
+  }
+  bool operator==(const SCCIterator &other) const {
+    return other.sccOp == this->sccOp && other.iterator == this->iterator;
+  }
+  bool operator!=(const SCCIterator &other) const { return !(*this == other); }
+  SCCNode operator*() {
+    Operation *op = iterator.getUser();
+    return SCCNode(op, sccOp.filter);
+  }
+  SCCIterator &operator++() {
+    ++iterator;
+    if (iterator != itEnd) {
+      Operation *nextOp = iterator.getUser();
+      if (nextOp == nullptr || (sccOp.filter && sccOp.filter(nextOp)))
+        iterator = itEnd;
```

Comment:
Is it intentional to skip an iterator to the end when there is  user which `filter` returns true? Isn't it necessary to set to the next user?